apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/Interval.cs

```diff
@@ -0,0 +1,381 @@
+    /// </param>
+    public static bool TryParse(string text, out Interval interval)
+    {
+        try
+        {
+            interval = Parse(text);
+            return true;
+        }
+        catch (Exception)
+        {
+            interval = null;
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Returns the ISO8601 string representation of this Interval.
+    /// </summary>
+    public override string ToString()
+    {
+        if (Equals(Zero))
+        {
+            return "P0Y";
+        }
+
+        int years = Months / MonthsInAYear ;
+        int months = Months % MonthsInAYear;
+
+        int hours = (int)(Nanoseconds / NanosecondsInAnHour);
+        long remainingNanoseconds = (long)(Nanoseconds % NanosecondsInAnHour);
+
+        long minutes = remainingNanoseconds / NanosecondsInAMinute;
+        remainingNanoseconds %= NanosecondsInAMinute;
+
+        decimal seconds = (decimal)remainingNanoseconds / NanosecondsInASecond;
+
+        StringBuilder intervalBuilder = new StringBuilder("P");
+
+        if (years != 0)
+        {
+            intervalBuilder.Append(years.ToString(CultureInfo.InvariantCulture));
+            intervalBuilder.Append("Y");
+        }
+
+        if (months != 0)
+        {
+            intervalBuilder.Append(months.ToString(CultureInfo.InvariantCulture));
+            intervalBuilder.Append("M");
+        }
+
+        if (Days != 0)
+        {
+            intervalBuilder.Append(Days.ToString(CultureInfo.InvariantCulture));
+            intervalBuilder.Append("D");
+        }
+
+        if (Nanoseconds != 0) // Has a time component
+        {
+            intervalBuilder.Append("T");
+
+            if (hours != 0)
+            {
+                intervalBuilder.Append(hours.ToString(CultureInfo.InvariantCulture));
+                intervalBuilder.Append("H");
+            }
+
+            if (minutes != 0)
+            {
+                intervalBuilder.Append(minutes.ToString(CultureInfo.InvariantCulture));
+                intervalBuilder.Append("M");
+            }
+
+            if (seconds != 0)
+            {
+                intervalBuilder.Append(seconds.ToString(CultureInfo.InvariantCulture));
+                intervalBuilder.Append("S");
+            }
+        }
+
+        return intervalBuilder.ToString();
+    }
+
+    /// <inheritdocs/>
+    public override bool Equals(object obj) =>
+        obj is Interval other
+        && other.Months == Months
+        && other.Days == Days
+        && other.Nanoseconds == Nanoseconds;
+
+    /// <inheritdocs/>
+    public override int GetHashCode() =>
+        GaxEqualityHelpers.CombineHashCodes(Days.GetHashCode(), Months.GetHashCode(), Nanoseconds.GetHashCode());
+
+    private static BigInteger SecondsToNanoseconds(decimal seconds)
+    {
+        // We only accept 9 decimal places.
+        decimal rounded = Math.Round(seconds, 9);
+        if (seconds != rounded)
```

Comment:
This will allow more than 9 decimal places to be specified, so long as they're zero. Do we want that? (I think it might give the wrong impression.)